File: server/src/main/com/broll/gainea/server/core/battle/FightingPower.java
package com.broll.gainea.server.core.battle;


import com.broll.gainea.misc.RandomUtils;
import com.broll.gainea.server.core.objects.Unit;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class FightingPower {

    private int diceCount;
    private int lowestNumber = 1;
    private int highestNumber = 6;
    private int numberPlus = 0;

    public FightingPower(Unit unit) {
        this(unit.getPower().getValue());
    }

    public FightingPower(int diceCount) {
        this.diceCount = diceCount;
    }

    public FightingPower changeDiceNumber(int count) {
        diceCount += count;
        if (diceCount < 1) {
            diceCount = 1;
        }
        return this;
    }

    public FightingPower changeHighestNumber(int delta) {
        this.highestNumber += delta;
        return this;
    }

    public FightingPower changeLowestNumber(int delta) {
        this.lowestNumber += delta;
        return this;
    }

    public FightingPower changeNumberPlus(int delta) {
        this.numberPlus += delta;
        return this;
    }

    public FightingPower setDiceCount(int diceCount) {
        this.diceCount = diceCount;
        if (diceCount < 1) {
            diceCount = 1;
        }
        return this;
    }

    public FightingPower withHighestNumber(int highestNumber) {
        this.highestNumber = highestNumber;
        return this;
    }

    public FightingPower withLowestNumber(int lowestNumber) {
        this.lowestNumber = lowestNumber;
        return this;
    }

    public FightingPower setNumberPlus(int numberPlus) {
        this.numberPlus = numberPlus;
        return this;
    }

    public int getDiceCount() {
        return diceCount;
    }

    public List<Integer> roll() {
        List<Integer> rolls = new ArrayList<>();
        for (int i = 0; i < diceCount; i++) {
            rolls.add(rollDice());
        }
        //sort descending
        Collections.sort(rolls, Collections.reverseOrder());
        return rolls;
    }

    private int rollDice() {
        return RandomUtils.random(lowestNumber, highestNumber) + numberPlus;
    }

    public int getHighestNumber() {
        return highestNumber;
    }

    public int getLowestNumber() {
        return lowestNumber;
    }

    public int getNumberPlus() {
        return numberPlus;
    }
}

File: server/src/main/com/broll/gainea/server/core/battle/FightingPower.kt
package com.broll.gainea.server.core.battle

import com.broll.gainea.misc.RandomUtils
import com.broll.gainea.server.core.objects.Unit
import java.util.Collections

class FightingPower(var diceCount: Int) {
    var lowestNumber = 1
        private set
    var highestNumber = 6
        private set
    var numberPlus = 0
        private set

    constructor(unit: Unit) : this(unit.power.getValue())

    fun changeDiceNumber(count: Int): FightingPower {
        diceCount += count
        if (diceCount < 1) {
            diceCount = 1
        }
        return this
    }

    fun changeHighestNumber(delta: Int): FightingPower {
        highestNumber += delta
        return this
    }

    fun changeLowestNumber(delta: Int): FightingPower {
        lowestNumber += delta
        return this
    }

    fun changeNumberPlus(delta: Int): FightingPower {
        numberPlus += delta
        return this
    }

    fun setDiceCount(diceCount: Int): FightingPower {
        var diceCount = diceCount
        this.diceCount = diceCount
        if (diceCount < 1) {
            diceCount = 1
        }
        return this
    }

    fun withHighestNumber(highestNumber: Int): FightingPower {
        this.highestNumber = highestNumber
        return this
    }

    fun withLowestNumber(lowestNumber: Int): FightingPower {
        this.lowestNumber = lowestNumber
        return this
    }

    fun setNumberPlus(numberPlus: Int): FightingPower {
        this.numberPlus = numberPlus
        return this
    }

    fun roll(): List<Int> {
        val rolls: MutableList<Int> = ArrayList()
        for (i in 0 until diceCount) {
            rolls.add(rollDice())
        }
        //sort descending
        Collections.sort(rolls, Collections.reverseOrder())
        return rolls
    }

    private fun rollDice(): Int {
        return RandomUtils.random(lowestNumber, highestNumber) + numberPlus
    }
}
